compiler/infra/Serializer.hpp

```diff
@@ -0,0 +1,281 @@
+ *
+ * This program and the accompanying materials are made available under
+ * the terms of the Eclipse Public License 2.0 which accompanies this
+ * distribution and is available at http://eclipse.org/legal/epl-2.0
+ * or the Apache License, Version 2.0 which accompanies this distribution
+ * and is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the
+ * Eclipse Public License, v. 2.0 are satisfied: GNU General Public License,
+ * version 2 with the GNU Classpath Exception [1] and GNU General Public
+ * License, version 2 with the OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
+#ifndef SERIALIZER_HPP
+#define SERIALIZER_HPP
+
+#include <type_traits>
+#include "infra/Assert.hpp"
+
+/**
+ * @class TR_Serializer
+ * @brief Provides facility to write and read trivially copyable data types to and from a memory buffer.
+ * When reading or writing, it maintains a pointer to the next location in the memory buffer where the data
+ * has to be read from or written to.
+ * It also has methods to compute the size of buffer required for writing data type.
+ * This class can be helpful when there is a requirement to serialize and deserialize objects by writing
+ * some specific fields of the class, or when the class is not trivially copyable.
+ */
+class TR_Serializer
+{
+   public:
+
+   TR_Serializer() : _buffer(0), _size(0), _bufferEnd(0), _current(0) {}
+
+   /**
+    * @brief Constructor that sets the memory buffer and the size of buffer used by the serializer
+    * @param [in] buffer pointer to memory buffer
+    * @param [in] size size of buffer in bytes
+    */
+   TR_Serializer(uint8_t *buffer, uint32_t size) : _buffer(buffer), _size(size), _bufferEnd(buffer + size), _current(buffer) {}
+
+   /**
+    * @brief adds the size of the parameter to the buffer size
+    * @tparam T type of the parameter
+    * @param [in] t size of this parameter is added to the buffer size
+    * @return void
+    */
+   template<typename T>
+   void addSize(T t)
+      {
+      static_assert(std::is_trivially_copyable<T>::value, "Cannot determine size of non-trivially copyable types");
+      _size += sizeof(T);
+      }
+
+   /**
+    * @brief adds the size of one or more parameters to the buffer size
+    * @tparam T type of the first parameter
+    * @tparam Args template parameter pack
+    * @param [in] t size of this parameter is added to the buffer size
+    * @param [in] args function parameter pack
+    * @return void
+    */
+   template<typename T, typename... Args>
```

Comment:
Same point as above about variadic templates; it's listed as unsupported in [1] but it states MSVC, so maybe it's not a problem on XLC?

[1] https://github.com/eclipse/omr/blob/master/doc/SupportedC%2B%2BFeatures.md#minimum-compiler-support